src/events/deposits.upsert.service.ts

```diff
@@ -288,4 +229,153 @@ export class DepositsUpsertService {
       { jobKey: 'sync_deposited_iron_to_telemetry', runAt },
     );
   }
+
+  async upsert(operation: UpsertDepositsOperationDto): Promise<Deposit[]> {
+    return this.prisma.$transaction((client) =>
+      this.upsertWithClient(client, operation),
+    );
+  }
+
+  async upsertWithClient(
+    prisma: BasePrismaClient,
+    operation: UpsertDepositsOperationDto,
+  ): Promise<Deposit[]> {
+    const networkVersion = this.config.get<number>('NETWORK_VERSION');
+    const blockHash = standardizeHash(operation.block.hash);
+
+    const shouldUpsertDeposit =
+      operation.type === BlockOperation.CONNECTED ||
+      operation.type === BlockOperation.DISCONNECTED;
+
+    let deposits = new Array<Deposit>();
+
+    if (shouldUpsertDeposit) {
+      for (const transaction of operation.transactions) {
+        if (operation.type === BlockOperation.CONNECTED) {
+          const amounts = new Map<string, number>();
+
+          for (const note of transaction.notes) {
+            if (note.memo) {
+              const amount = amounts.get(note.memo) ?? 0;
+              amounts.set(note.memo, amount + note.amount);
+            }
+          }
+
+          // Bulk load unique users and map by graffiti
+          const users = await this.usersService.findManyAndMapByGraffiti(
+            Array.from(amounts.keys()),
+          );
+
+          // Create deposit params for each deposit with a matching user
+          const depositParams = new Array<{
```

Comment:
Probably can use a client generated type here, maybe `DepositUpdateInput`

Reply:
Scratch this comment, I just tried importing and I get a type error due to some ugly codegening in the library, too bad.